builtin/providers/aws/utils.go

```diff
@@ -0,0 +1,13 @@
+package aws
+
+import "encoding/base64"
+
+// Base64Encode encodes data if the input isn't already encoded useing base64.StdEncoding.EncodeToString.
+// If the input is already base64 encoded, return the original input unchanged.
+func base64Encode(data []byte) string {
+	// Check whether the user_data is already Base64 encoded; don't double-encode
+	if _, base64DecodeError := base64.StdEncoding.DecodeString(string(data)); base64DecodeError != nil {
+		return base64.StdEncoding.EncodeToString(data)
```

Comment:
Since an error represents the expected path of not being encoded and therefore needing encoded, this may be clearer with a new more comments. `// not encoded, so encode it` `// already encoded`.

Or maybe even splitting out an `isBase64Encoded() bool` function?